### ISSUE TITLE
WriteSerializer returns Ok() on partial writes

### ISSUE THREAD

**Issue by Diomendius** [rkyv/rkyv#185]
```rust
use rkyv::{ser::Serializer, Archive, Serialize};

#[derive(Archive, Serialize)]
#[archive_attr(repr(C, packed))]
struct Foo {
    x: i32,
}

fn main() {
    let mut buf = [0_u8; std::mem::size_of::<Foo>() - 1];
    let mut ser = rkyv::ser::serializers::WriteSerializer::new(&mut buf[..]);
    let foo = Foo { x: 100 };
    ser.serialize_value(&foo)
        .expect_err("Serialized to a too-small buffer without error!");
}
```

This code panics at the `expect_err`, but only 3 of 4 bytes were actually written to `buf`. This actually matches the behaviour of `std::io::Write`, in which context an `Ok(usize)` makes sense as a partial write is not an error as such, especially when the underlying write mechanism is something like the `write` syscall, but this does not make sense in the context of rkyv's serializers as rkyv provides no way (as far as I can tell) to continue writing the last byte of `ArchivedFoo` after the first three have already been successfully written. In fact, `serialize_value()` returns an `Ok(usize)` that represents an entirely different concept than the `Ok(usize)` from `std::io::Result`, as it indicates the offset of the value just serialized, not the number of bytes written.

It's not obvious to me what the solution to this is, but as the value within `Ok()` is already used, `WriteSerializer::Error` needs to be a wrapper of some kind around a possible `std::io::Error`.

Maybe `WriteSerializer::Error` should be something like this:
```rust
enum Error {
    PartialWrite(usize),
    IoError(std::io::Error),
}
```

Or maybe some mechanism to retry writing the remaining bytes after a partial write should be provided. Probably the most ergonomic method here would be for `WriteSerializer` to keep internal state and provide some method to continue an interrupted write, but this might be complicated to implement as serialization isn't a simple `memcpy()`.

**Comment by djkoloski** [comment]
Good catch, this looks like a pretty easy fix. Internally, `WriteSerializer` is calling `write` but it looks like it should be calling `write_all` to have the expected behavior. A patch is on the way.

**Comment by djkoloski** [comment]
Fixed by ccee92c1a59a1839d03d7a6b4db5d8307d80195c and released in 0.7.9.
